### PR TITLE
Collect image dimensions from Europeana

### LINKED ISSUE #1484: Update Europeana to collect image dimensions
## Problem

We'd like to make sure that all of our image provider scripts are collecting data about image dimensions.

## Description

The Europeana script does not currently do so, but the information is available in the API via the [single record endpoint](https://pro.europeana.eu/page/record). We should update the provider script to collect this info.

There is precedent for making an additional API request per record, notably [Flickr](https://github.com/WordPress/openverse-catalog/blob/7bf8ec42408b1e4c76925ce159bdeec7de691c73/openverse_catalog/dags/providers/provider_api_scripts/flickr.py#L332) but also some others like Metropolitan and NYPL. That being said we should make sure that adding it here doesn't have any dramatic effect on performance.

## Additional context/dependent PRs

~Note that Europeana is currently turned off in production, because turning it on will initiate a backfill. We should wait for WordPress/openverse-catalog#644 to make sure that we can omit the DAG from Slack notifications, to prevent flooding the channel.~ <- This has been merged.

~We also have an issue to update Europeana to use the new API endpoint, and update any necessary fields: WordPress/openverse#1727. It may be helpful or even necessary to do this work first.~ <- https://github.com/WordPress/openverse-catalog/pull/974 has been merged.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by rwidom**
## Fixes
Fixes #1484 by @stacimc 

## Description

Adds image dimensions, filetype, and filesize from Europeana item data.

## Testing Instructions

At a high level, the goal is to run all the tests using `pytest`, then run the dag in airflow, then look at the data to confirm that it looks reasonable. 

- Run `just catalog/test-session` and then `pytest tests/dags/providers/provider_api_scripts/test_europeana.py`. 
- When I ran `pytest` alone from within the same test session, I got a bunch of errors from `test_ingestion_server.test_index_readiness_check` but that's just because the ingestion server stuff isn't started up when you're just running docker for a catalog test session.
- To get the local port for airflow (http://localhost:9090/home) working in docker, I ran `just catalog/recreate`. Then you can turn on the dag and it will go through daily runs and be sure to pick up some data. Check out the results using sql after `just catalog/pgcli`.

I temporarily replaced the dag start date with 8/8/2023 in provider_workflows.py, to get a smaller sample size for testing.  Only two days out of nine had data, but it seems like the 3 second timeout might be reasonable at least for the daily runs, if these numbers of records are comparable with production. I would need to know the total number of records in production to figure out if that is workable for a refresh.

```
*DAG*: `europeana_workflow`
*Date range*: 2023-08-08 -> 2023-08-09
*Duration of data pull tasks*: 24 mins
*Number of records upserted per media type*:
  - `image`: 474

*DAG*: `europeana_workflow`
*Date range*: 2023-08-15 -> 2023-08-16
*Duration of data pull tasks*: 56 mins, 20 secs
*Number of records upserted per media type*:
  - `image`: 1,113
```

The new fields are populated in all of the inserted records:

```
openledger> select filetype, count(*) records, count(filesize) filesize, count(width) width, count(height) height from image group by filetype order by 
 2 desc;
+----------+---------+----------+-------+--------+
| filetype | records | filesize | width | height |
|----------+---------+----------+-------+--------|
| jpg      | 1573    | 1573     | 1573  | 1573   |
| png      | 14      | 14       | 14    | 14     |
+----------+---------+----------+-------+--------+
SELECT 2
Time: 0.020s
```

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (obulat) [inline catalog/dags/providers/provider_api_scripts/europeana.py:275]: It is probably easier to check `if not item_response` in case there's another non-dictionary falsy value here

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (rwidom) [inline catalog/dags/providers/provider_api_scripts/europeana.py:267]: @obulat any thoughts on these delay settings/timing issues? Do you think additional performance testing would be valuable here?

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (rwidom) [reply catalog/dags/providers/provider_api_scripts/europeana.py:289]: Ahhh, fair enough! I think that `.startswith()` is probably more readable too, and it looks to be performant as long as we can trust that the `filetype` will be a string. Seem ok?

```
(openverse) % python -m timeit -s 'filetype="image/jpeg"' '(len(filetype)>=5 and filetype[:5]=="image")'
2000000 loops, best of 5: 130 nsec per loop
(openverse) % python -m timeit -s 'filetype="image/jpeg"' 'filetype.startswith("image")'                
2000000 loops, best of 5: 102 nsec per loop
```

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (obulat) [reply catalog/dags/providers/provider_api_scripts/europeana.py:267]: I don't know, but I think @stacimc might :)

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (obulat) [inline catalog/dags/providers/provider_api_scripts/europeana.py:179]: Is this delay correct?

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (obulat) [inline catalog/dags/providers/provider_api_scripts/europeana.py:283]: We could probably reduce nesting if we used something like
`if aggregations := item_response.get("object", {}).get("aggregations"):` and 
`for webresource in aggregation.get("webResources", [])`

**Review by obulat — APPROVED** [review]
I ran the script locally and the dimensions are being picked up well. I'll defer to @stacimc on the delay timing because she has more experience with it.
I added a comment on making the code a little bit less nested, but I don't want to block on it.

File: catalog/tests/dags/providers/provider_api_scripts/test_europeana.py
Comment (krysal) [inline catalog/tests/dags/providers/provider_api_scripts/test_europeana.py:122]: This looks like it could be extracted from the `item_full.json` file. Could it be the same for `ITEM_NOT_1ST_WEBRESOURCE`?

File: catalog/tests/dags/providers/provider_api_scripts/test_europeana.py
Comment (krysal) [inline catalog/tests/dags/providers/provider_api_scripts/test_europeana.py:280]: It would be nice to add the cases where there is only one of the dimensions.

**Review by krysal — COMMENTED** [review]
@rwidom Thanks for adding these four fields in one PR! :) Seems like this is changing the delay of 30 seconds that we currently have. @stacimc probably has better reasons to argue about that, as Olga said, but I'll just ask: did you confirm there is no way to grab these fields from the search endpoint?

**Comment by rwidom** [comment]
> @rwidom Thanks for adding these four fields in one PR! :) Seems like this is changing the delay of 30 seconds that we currently have. @stacimc probably has better reasons to argue about that, as Olga said, but I'll just ask: did you confirm there is no way to grab these fields from the search endpoint?

I looked in the search API documentation [here](https://pro.europeana.eu/page/search#metadata-sets), and it seems like it maybe should be possible (with some kind of profile magic?), but it's not clear to me how to do it. Definitely when I look at the test files we've actually downloaded, I don't see anything that looks like those fields. It would be great if we didn't have to query individual objects, because then we could leave the 30 seconds alone.

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (stacimc) [inline catalog/dags/providers/provider_api_scripts/europeana.py:283]: I'd need to double check, but I think a generator would be faster, and IMO a little easier to read (although opinions may vary 😄). Something like:

```
aggregations = item_response.get("object", {}).get("aggregations", [])
for aggregation in aggregations:
    return next(
        (resource for resource in aggregation.get("webResources", []) if resource.get("ebucoreHasMimeType", "").startswith("image")),
        {} # Returns `{}` by default
    )
```

**Review by stacimc — CHANGES_REQUESTED** [review]
Thanks for taking this on @rwidom!

I spent some time looking at the API documentation and found the aggregations very confusing. My takeaway is that what we're getting back from the search endpoint is a collection of "metadata records", each of which may have multiple actual images associated with it. When we query the individual record endpoint, then, we may get multiple images in the `aggregations` with different filesizes etc.

So I took a look at one of the examples for record id `/2021650/memorix_0000ee69_1b9e_6823_478f_c88af61736c6`. The image url that we get from ingestion is https://images.memorix.nl/nda/thumb/fullsize/010a31d7-9316-54d7-6ad0-83b9914688d9.jpg. That url _is_ one of the webresources we can find at the record endpoint, but it's not the first one -- that's a (smaller) version of the same image, https://images.memorix.nl/nda/thumb/640x480/010a31d7-9316-54d7-6ad0-83b9914688d9.jpg.

The result is that, at least some of the time, the final record that gets ingested will have image dimensions for a different size than the one we're linking out to, which I don't think we want. Moreover based on the API documentation, I'm worried that we might also get different photos entirely -- like related photos or multiple angles of the same object, in which case it would be really strange to get the dimensions of the wrong image.

I wonder if we could try to filter the webresources for one that exactly matches the url we used for `image_url`?

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (rwidom) [reply catalog/dags/providers/provider_api_scripts/europeana.py:248]: I hear that. I feel nervous about it, because the overall structure seems so flexible that I worry about things getting jumbled up that way. What would you think of something like `data | {"item webresource": webresource}`?

Comment (stacimc) [reply catalog/dags/providers/provider_api_scripts/europeana.py:248]: Good point, I think caution is understandable! I like that idea.

**Comment by rwidom** [comment]
@stacimc , I think I made the code-only changes you asked for, except the timing one. I'm not sure the best way to reach out to Europeana, but while I was looking, I came across [this "Harvesting and Downloads" page](https://pro.europeana.eu/page/harvesting-and-downloads) which seems directly relevant to us, though I haven't looked in more detail. Hmmm...

**Comment by rwidom** [comment]
Thanks for talking through the timeout issue @stacimc  and @AetherUnbound !

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (stacimc) [reply catalog/dags/providers/provider_api_scripts/europeana.py:277]: I think we can go ahead and remove this comment now!

File: catalog/dags/providers/provider_reingestion_workflows.py
Comment (stacimc) [inline catalog/dags/providers/provider_reingestion_workflows.py:59]: 🥳

File: catalog/dags/providers/provider_api_scripts/europeana.py
Comment (stacimc) [inline catalog/dags/providers/provider_api_scripts/europeana.py:248]: Could we change the key to just `webresource` (or `item_webresource` with the underscore)?

**Review by stacimc — COMMENTED** [review]
This looks great! Just a couple tiny clean-up requests and it looks like we need to `just catalog/generate-dag-docs`, but then this should be good to go! 🚢

**Review by stacimc — APPROVED** [review]
This all looks and tests great @rwidom! 🚀 Thanks for leading the discussion on this one, I'm excited for this :)

**Pull request merged**
